### PR TITLE
refactor strongTypePtr

### PR DESCRIPTION
**Pull request by zhaojuanmao**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32959 jit pickling rref
* **#33590 refactor strongTypePtr**

Differential Revision: [D20015521](https://our.internmc.facebook.com/intern/diff/D20015521/)

### CODE REVIEW

File: test/cpp/jit/CMakeLists.txt [hunk 1]
```diff
@@ -27,6 +27,8 @@ elseif (USE_ROCM)
 
   target_link_libraries(test_jit PRIVATE caffe2_gpu)
 
+  target_link_libraries(test_jit PRIVATE torch_cpu)
```
Comment (suo) [inline test/cpp/jit/CMakeLists.txt]: What is the purpose of this?

File: aten/src/ATen/core/ivalue.cpp [hunk 2]
```diff
@@ -304,6 +307,17 @@ std::vector<std::pair<IValue, IValue>> iterationOrder(const c10::Dict<IValue, IV
...
 }
 
+StrongTypePtr::StrongTypePtr(
+    std::shared_ptr<torch::jit::script::CompilationUnit> cu,
+    std::shared_ptr<Type> type) {
+  cu_ = std::move(cu);
+  type_ = type;
+  TORCH_INTERNAL_ASSERT(type_);
+  if (type_->cast<ClassType>()) {
+    TORCH_INTERNAL_ASSERT(cu_);
```
Comment (suo) [inline aten/src/ATen/core/ivalue.cpp]: Can we add a more detailed error message in this case? Also, one useful check might be to assert that the class type's owning compilation unit is the same as the `cu` we passed in here.

File: aten/src/ATen/core/ivalue.h [hunk 3]
```diff
@@ -656,13 +656,10 @@ struct CAFFE2_API WeakIValue final {
 struct StrongTypePtr {
```
Comment (suo) [inline aten/src/ATen/core/ivalue.h]: Can you update the comment for this struct?

**Review by suo — APPROVED** [review]

File: test/cpp/jit/CMakeLists.txt (hunk 1 above)
Comment (zhaojuanmao) [reply test/cpp/jit/CMakeLists.txt]: it is trying to resolve link error, but it seems not to resolve the failures in CI

File: aten/src/ATen/core/ivalue.cpp (hunk 2 above)
Comment (zhaojuanmao) [reply aten/src/ATen/core/ivalue.cpp]: will do

File: aten/src/ATen/core/ivalue.h (hunk 3 above)
Comment (zhaojuanmao) [reply aten/src/ATen/core/ivalue.h]: will do

**Pull request closed without merging**